### PR TITLE
b-bit integer circuit 

### PR DESCRIPTION
**Pull request by degregat**
@henrycg This is taking into account the comments from #89. 

I added some convenience functions to ease the creation of a compatible config without changing the underlying data structures and algorithms.

### CODE REVIEW

**Comment by henrycg** [comment]
Very cool! Thanks for the hard work on this! The revised PR looks very clean. I'll do a more careful review over the next few days.

**Comment by degregat** [comment]
Awesome! It wasn't too bad since all the code was written already. Just needed to figure out a sensible API. This was the smallest extension that I thought mades sense, so if you see some possible improvements on that front, please say so. Also, this code assumes out of band sync of precision between client and server parts, I hope that is clear enough from the documentation.

File: include/mprio.h [hunk 1]
```diff
@@ -93,14 +93,36 @@ void Prio_clear();
...
+ */
+PrioConfig PrioConfig_new_int(int nInts, int prec, PublicKey serverA,
+                              PublicKey serverB, const unsigned char* batchId,
+                              unsigned int batchIdLen);
+
 void PrioConfig_clear(PrioConfig cfg);
 int PrioConfig_numDataFields(const_PrioConfig cfg);
 
+/*
+ * Return the number of integers with precision prec.
```
Comment (henrycg) [inline include/mprio.h:112]: Maybe instead "Return the number of integers with precision prec that this configuration supports."

File: prio/config.c [hunk 2]
```diff
@@ -22,6 +22,12 @@ PrioConfig_maxDataFields(void)
   return (n_roots >> 1) - 1;
 }
 
+int
+PrioConfig_maxIntEntries(int prec)
```
Comment (henrycg) [inline prio/config.c:26]: Since the input here is a signed `int`, do we want to check whether `prec <= 0` and do something sensible if so?

File: prio/config.c [hunk 3]
```diff
@@ -67,6 +73,15 @@ PrioConfig_new(int n_fields, PublicKey server_a, PublicKey server_b,
   return cfg;
 }
 
+PrioConfig
+PrioConfig_new_int(int num_ints, int prec, PublicKey server_a,
+                   PublicKey server_b, const unsigned char* batch_id,
+                   unsigned int batch_id_len)
+{
+  return PrioConfig_new(num_ints * prec, server_a, server_b, batch_id,
```
Comment (henrycg) [inline prio/config.c:81]: I think there could be an integer overflow problem here if `num_ints * prec > `INT_MAX`. I think that one way to handle this safely would be to first check that `num_ints <= PrioConfig_maxIntEntries(prec)`.

File: prio/config.c [hunk 4]
```diff
@@ -92,6 +107,12 @@ PrioConfig_numDataFields(const_PrioConfig cfg)
   return cfg->num_data_fields;
 }
 
+int
+PrioConfig_numIntEntries(const_PrioConfig cfg, int prec)
+{
+  return cfg->num_data_fields / prec;
```
Comment (henrycg) [inline prio/config.c:113]: As above, we should handle the case `prec <= 0` gracefully here.

File: prio/client.c [hunk 5]
```diff
@@ -336,6 +336,64 @@ PrioClient_encode(const_PrioConfig cfg, const bool* data_in,
   return rv;
 }
 
+// Encode long as bools and write them to an array with offset.
+SECStatus
+long_to_bool(bool* dst, long x, int prec, int entry)
```
Comment (henrycg) [inline prio/client.c:341]: Could you add some unit tests for this function? This is probably the most subtle part of the PR, so it'd be nice to get a bit more confidence that it's bug-free.

One hacky (but functional) way to test non-public functions is to just `#include "prio/client.c" in the test.

File: prio/client.c [hunk 6]
```diff
@@ -336,6 +336,64 @@ PrioClient_encode(const_PrioConfig cfg, const bool* data_in,
...
+PrioClient_encode_int(const_PrioConfig cfg, const int prec,
+                      const long* data_int, unsigned char** forServerA,
+                      unsigned int* aLen, unsigned char** forServerB,
+                      unsigned int* bLen)
+{
+  SECStatus rv = SECSuccess;
+  bool* data_bool = NULL;
+
+  int num_ints = PrioConfig_numIntEntries(cfg, prec);
+  P_CHECKCB(cfg->num_data_fields == prec * num_ints);
```
Comment (henrycg) [inline prio/client.c:380]: I was thinking that there might be an integer-overflow issue here, but on second look, it's probably fine.

File: prio/config.c (hunk 2 above)
Comment (degregat) [reply prio/config.c:26]: You're right. That check was in encode.c of the old PR and I forgot to port it over. There I just did P_CHECKCB(), but if you point me to an example of error handling, I can add that there.

File: prio/client.c (hunk 5 above)
Comment (degregat) [reply prio/client.c:341]: Will do.

File: prio/client.c (hunk 6 above)
Comment (degregat) [reply prio/client.c:380]: As long as PrioConfig_maxDataFields() * prec <= INT_MAX we are fine. Does that always hold?

If so, this check (together with (MP_DIGIT_MAX > ULLONG_MAX)) will prevent users from trying to aggregate to many bits/ints.

File: prio/config.c (hunk 3 above)
Comment (degregat) [reply prio/config.c:81]: The old PR does ```P_CHECKCB(cfg->precision <= 32); P_CHECKCB(cfg->precision >= 1);```. Do we want to further constrain the precision or just prevent overflows?

Comment (henrycg) [reply prio/config.c:81]: Either way works for me. I can't think of any realistic applications for using integer values with >32 bits of precision, but it seems fine to allow this in case someone has a reason to do this.

File: prio/config.c (hunk 2 above)
Comment (henrycg) [reply prio/config.c:26]: Since this function returns an `int` instead of an error code, you could just return `0` if `prec <= 0`.

File: prio/client.c (hunk 6 above)
Comment (henrycg) [reply prio/client.c:380]: > As long as "PrioConfig_maxDataFields() * prec <= INT_MAX" we are fine. Does that always hold?

I don't think that the above relation always holds, but I think that `PrioConfig_numIntEntries() * prec <= INT_MAX` probably does hold.

Either way, it might be good to add an assertion just to make sure that there isn't an integer overflow. Alternatively, we could put a global cap on `prec` everywhere (to something like 32, 64, or 128) just to be very sure that we will never even get close to an overflow on a 32- or 64-bit platform. Either way works for me.

File: prio/client.c (hunk 5 above)
Comment (degregat) [reply prio/client.c:341]: Tests are done, tell me if you think they are sufficient.

File: prio/client.c (hunk 6 above)
Comment (degregat) [reply prio/client.c:386]: BBIT_MAX_PREC will usually be 63 now, since this is what fits in longs on most platforms. We could use unsigned long, but then I will need to cover different edge cases in the encoding, since the compiler optimizes some checks away. We could also use long long?

**Comment by degregat** [comment]
Running tests takes a long time now. Should I reduce the amount of entries in some of them, or run some with lower precision?

**Comment by degregat** [comment]
If you're happy with this, a squash merge for a concise feature commit would be great, but if you want to keep the history thats fine too. I still need to get used to using a CI and the formatter :)

**Comment by degregat** [comment]
Another thing that came to mind: If we want to support negative numbers as well in the future, with another set of API calls, we might want to rename the suffixes hers to _uint.

File: prio/client.c (hunk 6 above)
Comment (henrycg) [reply prio/client.c:386]: Max'ing out at 63 bits sounds fine to me. That should handle all of the use cases that we have discussed.

**Comment by henrycg** [comment]
Good point. Renaming to _uint might make sense.

Let me know when you'd like me to review the last batch of edits. I will have time to look through it all in the second half of next week.

File: prio/client.c (hunk 6 above)
Comment (degregat) [reply prio/client.c]: Missed that the results can hold at most an MP_DIGIT which is at most ULONG_MAX. I set BBIT_PREC_MAX down to 42, so we will be able to aggregate (2^64-1)/(2^42-1) =~ 4194304 submissions.

File: prio/client.c (hunk 5 above)
Comment (degregat) [reply prio/client.c:341]: One could also change this to use unsigned long, so there would be one more compiler check. On the other hand, we will feed it output from fixed point and additively shifted negative number functions, so we would likely need to do type conversions or casting, so doing range checks P_CHECKCB on longs seems preferable. Also it is using long now, and does the correct thing, so we would not need to change it. What do you think?

**Comment by degregat** [comment]
This should be it! There are two last comments on the code which you might want to look at, but they are mainly clarifications.

**Comment by henrycg** [comment]
Excellent. Thanks! I have a deadline midweek this week, but I will plan to review this on Friday.

File: prio/client.c (hunk 6 above)
Comment (degregat) [reply prio/client.c]: Do you think thats a good choice here? If so, we can resolve this too.

**Comment by degregat** [comment]
Thanks for the detailed feedback!

File: prio/client.c (hunk 6 above)
Comment (henrycg) [reply prio/client.c]: Yes, I think that's fine.

**Comment by degregat** [comment]
The macOS build has failed, but Bionic and FreeBSD have passed.

**Comment by degregat** [comment]
The macOS build has failed/stalled, but Bionic and FreeBSD have passed: @rhelmer do you have a hint what I need to do?

**Comment by henrycg** [comment]
Once I rebase this PR onto master (so to include commit 19faab708387) I'm able to compile and pass all the tests on a MacOS machine.

So it looks like the problem may just be that Travis CI is timing out after 10 minutes of no log output. One simple fix, which we probably should do anyways, would be to have Travis CI invoke the tests as `build/ptest/ptest -vvv` so that we print a line before each test runs.

**Comment by degregat** [comment]
See #96

**Comment by henrycg** [comment]
Fixed by #96.
